python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -608,6 +608,151 @@ def _scan(self, op: str, *args, **kwargs):
 
     aggregate = agg
 
+    def _head_tail(self, n, *, take_head=True):
+        """Return the head or tail of each group
+
+        Parameters
+        ----------
+        n
+           Number of entries to include (if negative, number of
+           entries to exclude)
+        take_head
+           Do we want the head or the tail of the group
+
+        Returns
+        -------
+        New DataFrame or Series
+
+        Notes
+        -----
+        Unlike pandas, this returns an object in group order, not
+        original order
+        """
+        # A more memory-efficient implementation would merge the take
+        # into the grouping, but that probably requires a new
+        # aggregation scheme in libcudf. This is probably "fast
+        # enough" for most reasonable input sizes.
+        _, offsets, _, group_values = self._grouped()
```

Comment:
Note that this OOMs for setups where I _really_ think it shouldn't. I tried (on an A6000, 48GiB device memory) creating a dataframe with two int-columns with 10^9 rows (and 10^6 groups) [8GiB data]. `df.groupby("a")._grouped()` consistently OOMs for me. AIUI this is a sort-based implementation of reordering which I would have thought runs in-place on a copy of the table, so I was expecting to peak at around 16GiB of data (even if it is out-of-place I would expect not much more than 24GiB of peak footprint). That said, I don't know enough about the libcudf implementation of groupbys to understand if my expectation is reasonable.